src/transformStateWithClones.js

```diff
@@ -7,7 +7,52 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const statesHistory = [];
+  let newState = state;
+
+  for (const action of actions) {
+    newState = { ...newState };
+
+    switch (action.type) {
+      case 'addProperties':
+        newState = addProperties(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        newState = removeProperties(newState, action.keysToRemove);
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
+
+    statesHistory.push(newState);
+  }
+
+  return statesHistory;
+}
+
+/**
+ * @param {Object} state
+ * @param {string[]} keysToAdd
+ *
+ * @return {Object}
+ */
+function addProperties(state, keysToAdd) {
+  return Object.assign(state, keysToAdd);
+}
+
+/**
+ * @param {Object} state
+ * @param {string[]} keysToRemove
+ *
+ * @return {Object}
+ */
+function removeProperties(state, keysToRemove) {
```

Comment:
[CODE STYLE] - don't mutate object or arrays. The `removeProperties` function should return a new object instead of mutating the passed `state` object. Use a method that doesn't mutate the original object, like creating a new object with the properties that should remain.